File: src/user_operation.c
#include "../include/operations.h"
#include "../include/session.h"
#include "../include/user.h"
#include <string.h>
#include <stdio.h>
#include <arpa/inet.h>

void handle_user_operation(int client_sock, uint8_t *buffer) {
    printf("Buffer contents:\n");
    for (size_t i = 0; i < 32; i++) {
        printf("%02x ", buffer[i]);
        if ((i + 1) % 16 == 0) {
            printf("\n");
        }
    }
    printf("\n");

    uint8_t opcode = buffer[0]; // Extract the opcode
    uint8_t flag = buffer[1]; // Extract the user operation flag
    uint16_t username_len = ntohs(*(uint16_t*)&buffer[4]);
    uint16_t password_len = ntohs(*(uint16_t*)&buffer[6]);

    printf("Opcode: %u\n", opcode);
    printf("Flag: %u\n", flag);
    printf("Extracted username length: %u\n", username_len);
    printf("Extracted password length: %u\n", password_len);

    if (username_len > USERNAME_LENGTH || password_len > PASSWORD_LENGTH) {
        printf("Invalid username or password length\n");
        send_response_code(client_sock, 0xFF, 0);  // Failure
        return;
    }

    char username[username_len + 1];
    char password[password_len + 1];

    memcpy(username, &buffer[12], username_len);
    memcpy(password, &buffer[12 + username_len], password_len);

    username[username_len] = '\0';
    password[password_len] = '\0';

    printf("Username: %s\n", username);
    printf("Password: %s\n", password);

    if (opcode == 0x01) { // Assuming 0x01 is the opcode for user operations
        if (flag == 0x00) {
            // Handle login operation
            User* user = authenticate_user(username, password);
            if (user != NULL) {
                Session* session = create_session(user->role);
                if (session) {
                    send_response_code(client_sock, 0x01, session->session_id);
                    printf("Login successful, session ID: %u\n", session->session_id);
                } else {
                    send_response_code(client_sock, 0xFF, 0);
                    printf("Failed to create session\n");
                }
            } else {
                send_response_code(client_sock, 0xFF, 0);  // Failure
                printf("Invalid credentials\n");
            }
        } else if (flag == 0x01 || flag == 0x02 || flag == 0x03) {
            // Handle create user operation
            UserRole role = READ;
            if (flag == 0x02) {
                role = READ_WRITE;
            } else if (flag == 0x03) {
                role = ADMIN;
            }

            int result = register_user(username, password, role);
            if (result == 0) {
                send_response_code(client_sock, 0x01, 0);  // Success
                printf("User created successfully\n");
            } else {
                send_response_code(client_sock, 0xFF, 0);  // Failure
                printf("Failed to create user\n");
            }
        } else if (flag == 0xff) {
            if (delete_user(username) == 0) {
                // Send success response
                send_response_code(client_sock, 0x00, 0);  // Success
                printf("User deleted successfully\n");
            } else {
                // Send failure response
                send_response_code(client_sock, 0xFF, 0);  // Failure
                printf("User deletion failed\n");
            }
        } else {
            // Handle other operations, like delete user
            printf("Unknown operation flag: %u\n", flag);
            send_response_code(client_sock, 0xFF, 0);  // Failure
        }
    } else {
        // Handle other opcodes
        printf("Unknown opcode: %u\n", opcode);
        send_response_code(client_sock, 0xFF, 0);  // Failure
    }
}

File: client/operations/user_operation.py
import struct
from operations.util.session import set_session_id


def handle_user_operation(sock, session_id):
    print("User Operations:")
    print("1: Login")
    print("2: Create User")
    print("3: Delete User")
    user_choice = input("Enter operation number: ")

    if user_choice == "1":
        username = input("Enter username: ")
        password = input("Enter password: ")
        session_id = send_login_request(sock, username, password)
        if session_id is None:
            print("Exiting due to login failure.")
            return
        set_session_id(session_id)

    elif user_choice == "2":
        print("Creating user operations:")
        print("Select level to create user")
        print("1: Create the given user with Read permissions")
        print("2: Create the given user with Read/Write permissions")
        print("3: Create the given user with Admin permissions")
        choice = input("Enter level of operation: ")
        if choice in ["1", "2", "3"]:
            permissions = {
                "1": 0x01,  # Read
                "2": 0x03,  # Read/Write
                "3": 0x04  # Admin
            }
            username = input("Enter username: ")
            password = input("Enter password: ")
            send_create_user_request(sock, session_id, username, password, permissions[choice])

    elif user_choice == "3":
        username = input("Enter username: ")
        send_delete_user_request(sock, session_id, username)
    else:
        print("Invalid operation.")


def print_buffer(buffer):
    print("Buffer to be sent:")
    print(" ".join(f"{b:02x}" for b in buffer))


def send_login_request(sock, username, password):
    username_bytes = username.encode('utf-8')
    password_bytes = password.encode('utf-8')

    session_id_bytes = b'\x00\x00\x00\x00'

    # Pack the request
    request_format = f'>B B 2x H H 4s {len(username_bytes)}s {len(password_bytes)}s'
    request = struct.pack(request_format, 0x01, 0x00, len(username_bytes), len(password_bytes), session_id_bytes,
                          username_bytes, password_bytes)

    # Print the buffer before sending
    print_buffer(request)

    # Send request
    sock.sendall(request)

    # Handle response from server
    response = sock.recv(8)
    return_code, reserved, session_id = struct.unpack('>B B 4s 2x', response)
    session_id = int.from_bytes(session_id, byteorder='big')

    if return_code == 0x01:
        print("Login successful.")
        set_session_id(session_id)
        return session_id
    else:
        print("Login failed.")
        return None


def send_create_user_request(sock, session_id, username, password, flag):
    print(flag)
    username_bytes = username.encode('utf-8')
    password_bytes = password.encode('utf-8')

    request_format = f'>B B 2x H H 4s {len(username_bytes)}s {len(password_bytes)}s'
    request = struct.pack(request_format, 0x01, flag, len(username_bytes), len(password_bytes)
                          , session_id.to_bytes(4, 'big'), username_bytes, password_bytes)

    # Print the buffer before sending
    print_buffer(request)

    # Send request
    sock.sendall(request)

    # Handle response from server
    response = sock.recv(1024)
    if response[0] == 0x01:
        print("User created successfully.")
    else:
        print("User creation failed.")


def send_delete_user_request(sock, session_id, username):
    username_bytes = username.encode('utf-8')
    password_bytes = b''  # Empty password for deletion
    password_len = 0

    request_format = f'>B B 2x H H 4s {len(username_bytes)}s {len(password_bytes)}s'
    # Create the request buffer
    request = struct.pack(request_format, 0x01, 0xff, len(username_bytes), len(password_bytes)
                          , session_id.to_bytes(4, 'big'), username_bytes, password_bytes)

    # Print the buffer before sending
    print_buffer(request)

    # Send request
    sock.sendall(request)

    # Handle response from server
    response = sock.recv(1024)
    if response[0] == 0x00:
        print("User deleted successfully.")
    else:
        print("User deletion failed.")
